### PR TITLE
Use Google Sheets API v4

### LINKED ISSUE #16210: Google sheets data import
#### Expected behaviour
Importing with google sheets loads the data.

#### Actual behaviour
The google endpoint being used in Highcharts has started to sometimes returns a 404. The issue is pretty sporadic though, sometimes it pulls the data fine, other times it returns a 404. More information about this can be found here:

https://support.google.com/docs/thread/121088347/retrieving-data-from-sheets-results-in-404-error-50-of-the-time?hl=en

I know this is a google issue, but it seems to be affecting all sheet charts created through the Highcharts library now.

#### Live demo with steps to reproduce
You should be able to see this with all spreadsheet charts now, I've linked your demo below:
https://jsfiddle.net/gh/get/library/pure/highcharts/highcharts/tree/master/samples/highcharts/data/google-spreadsheet/

It may work when you load it first, but if you click the run button a few times, it breaks. The behaviour is inconsistent.

### PR DESCRIPTION
**Pull request by TorsteinHonsi**
Added internal support for Google Sheets API v4, fixing stability problems with loading data from spreadsheets. 

---

#### Upgrade note
Existing setups with Google Sheets stopped working in August 2021 as Google phased version 3 of their Google Sheets API. Highcharts has now been upgraded to version 4 of the API, but this will not work unless an API key is given in the `data.googleAPIKey` option. See the [comprehensive tutorial](https://handsondataviz.org/google-sheets-api-key.html) from Hands-On Data Visualization on how to obtain the key. Additionally, `data.googleSpreadsheetWorksheet` is deprecated and replaced by `data.googleSpreadsheetRange`, which allows for more detailed instructions on what data to load.

#### Live demo
https://jsfiddle.net/highcharts/c34omvha/

#### To do
 - [x] Make it work with basic setup
 - [x] Intelligent handling of datetime columns
 - [x] Migrate `startRow`, `endRow`, `startColumn` and `endColumn` options.
 - [x] Documentation on how to set up the API key. Handsondataviz has a [comprehensive tutorial](https://handsondataviz.org/google-sheets-api-key.html), perhaps just refer to that.
 - [x] API docs, `googleSpreadsheetKey` and linked resources has some outdated information.

Fixes #16210

### CODE REVIEW

**Comment by jsilff** [comment]
Rather than getRange(), please allow passing of a worksheet name so that you can process data specific worksheets within the selected spreadsheet (options.googleSpreadsheetWorksheet).

**Comment by TorsteinHonsi** [comment]
> Rather than getRange(), please allow passing of a worksheet name so that you can process data specific worksheets within the selected spreadsheet (options.googleSpreadsheetWorksheet).

@jsilff You can do that through the new `googleSpreadsheetRange` option: https://jsfiddle.net/highcharts/c34omvha/2/ . I propose to deprecate the `googleSpreadsheetWorksheet` option because it is limited to the internal worksheet keys, so it wouldn't be backwards compatible if we passed it directly to the `values`/`range` parameter in Google Sheets API v4. Also, the new option is more flexible as it lets you define both the sheet and the range of cells to load.

https://github.com/highcharts/highcharts/blob/bfd42d35daed665b82511c5029a3c759205df9c6/ts/Extensions/Data.ts#L481-L500

File: ts/Extensions/Data.ts [hunk 1]
```diff
@@ -1749,38 +1788,53 @@ class Data {
...
-            // use sheet 1 as the default rather than od6
-            // as the latter sometimes cause issues (it looks like it can
-            // be renamed in some cases, ref. a fogbugz case).
-            worksheet = options.googleSpreadsheetWorksheet || 1,
-            startRow = options.startRow || 0,
-            endRow = options.endRow || Number.MAX_VALUE,
-            startColumn = options.startColumn || 0,
-            endColumn = options.endColumn || Number.MAX_VALUE,
-            refreshRate = (options.dataRefreshRate || 2) * 1000;
+            refreshRate = Math.min((options.dataRefreshRate || 2) * 1000, 4000);
```
Comment (KacperMadej) [inline ts/Extensions/Data.ts:1795]: If this should run like the old code then `Math.max` should be used.

File: ts/Extensions/Data.ts [hunk 2]
```diff
@@ -1749,38 +1788,53 @@ class Data {
...
+            }
+
+            const alphabet = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ';
+            const start = (alphabet.charAt(options.startColumn || 0) || 'A') +
+                ((options.startRow || 0) + 1);
+
+            let end = alphabet.charAt(pick(options.endColumn, -1)) || 'ZZ';
+            if (defined(options.endRow)) {
+                end += options.endRow + 1;
+            }
```
Comment (KacperMadej) [inline ts/Extensions/Data.ts:1813]: `String.fromCharCode(65 + x)` or similar could be used to eliminate the `alphabet` const.

File: ts/Extensions/Data.ts [hunk 3]
```diff
@@ -459,13 +479,32 @@ declare global {
...
+ * [googleSpreadsheetKey](#data.googleSpreadsheetKey). See
+ * [developers.google.com](https://developers.google.com/sheets/api/reference/rest/v4/spreadsheets.values/get)
+ * for details.
+ *
+ * If given, it takes precedence over `startColumn`, `endColumn`, `startRow` and
+ * `endRow`.
+ *
+ * @example
+ * googleSpreadsheetRange: 'Fruit Consumption' // Load a named worksheet
+ * googleSpreadsheetRange: 'A-Z' // Load columns A to Z
```
Comment (KacperMadej) [inline ts/Extensions/Data.ts:492]: `-` is not working, should be `A:Z`

File: ts/Extensions/Data.ts (hunk 2 above)
Comment (TorsteinHonsi) [reply ts/Extensions/Data.ts:1813]: It will fail if `startColumn` > 26 though... With the current code, that case will fall back to 'A'

**Review by KacperMadej — APPROVED** [review]

**Comment by merekat** [comment]
Your jsfiddles do not work due to 400 errors.

**Comment by KacperMadej** [comment]
@merekat Which ones? The changes done in the PR will be published after the next release.

**Comment by oliverater** [comment]
@KacperMadej @TorsteinHonsi thanks for the quick fix. Any idea when we can expect the next release?

**Comment by paquetden** [comment]
Hello, I was planning on using this quick fix [https://jsfiddle.net/highcharts/c34omvha/2/](url) that was working on friday. Can you make the files availables? It's no longer working. Thank you very much for your help!

**Comment by HP4k1h5** [comment]
@paquetden i made adapted this fiddle to use a recent commit sha
https://jsfiddle.net/1k830ojs/1/

**Comment by paquetden** [comment]
> @paquetden i adapted this fiddle to use a recent commit sha
> https://jsfiddle.net/1k830ojs/1/

Thank you!

**Comment by KacperMadej** [comment]
In general after a merge into the master branch and deletion of the merged branch, a demo that is based on the master code should work fine before the release: https://jsfiddle.net/BlackLabel/8cktapse/

**Comment by TorsteinHonsi** [comment]
@paquetden Please remember that code from github.highcharts.com is not safe to use in production, so once the release is out, you should switch to code.highcharts.com.

**Comment by paquetden** [comment]
@TorsteinHonsi Thank you! I have loaded the files on our server and I will do the switch when the release is out.

**Comment by jsilff** [comment]
FYI, the new release was posted and it is working great! Thank you for your hard work @TorsteinHonsi!

**Pull request merged**